ARTEMIS-4259 JMS consumer + FQQN + selector not working

jbertram opened the pull request.

gemmellr commented:
More specifically, auto-created 'topic consumer + FQQN + filter' queues, and expecting that only messages matching that consumers filter will get to the created queue, and ultimately the consumer?

Looks like for Core, it didnt do any filtering at all for the auto-created FQQN+filter case, as neither queue creation or consumer had a filter applied. For AMQP, it looks like the consumers are filtered, but the queue wasnt and thus would get all messages on the address and retain any that dont match the consumer(s) filter(s). If a pre-created queue existed with an appropriate filter, it seems like it would have worked for both.

This change looks like it fixes the auto-creation case, though seems like it would introduce and/or retain some oddities for working with existing queues either pre-defined or previously auto-created (plus may even prevent some niche things that worked before).

For queues that already exist and that either have no filter or have a different filter than a new consumer uses, it doesnt seem like it will be noticed since nothing is done with the queue filter when the queue already exists. For Core that seems like it could mean if the existing queue has no filter it still does no filtering even when its desired it should , and for the mismatched-filters case it might actually filter the 'wrong thing' and again allow non-matching messages to get to the consumer, since it still looks to rely solely on the hopefully-existing-and-hopefully-matching queue filter alone, with no consumer filter put in place at all. For AMQP, it seems it could once again pile up any messages that dont match the new consumers filter for the no-existing-queue-filter case (since it will still get all messages but the consumer will be filtered), and in the mismatched filter case it may not get all messages it 'should' and may pile up unmatched messages again since the queue filter doesnt match the consumers, though at least no 'non-matching' messages would get to the consumer due to its filter (note it will often be double-filtering though since both queue and consumer can have filters).

Its not clear to me what can/should be done about that. Document it and ignore? Check if queue is filtered and throw if there is a mismatch (but that might actually be deliberate for a pre-defined queue filter..)? (For regular JMS durable subs the existing sub is to be seamlessly destroyed and a fresh sub begun when filter changes..same for shared subs, if not currently in use by others).

One thing is this could prevent some existing possible niche behaviours. Previously it seems you could perhaps have had multiple filtered AMQP consumers having different filters pulling from the same FQQN (say, structured such that all messages could be consumed by someone). This will break such things for auto-created usage, since the queue would now be filtered based only on the first/creating consumers filter.  JMS specifically doesnt allow such stuff with regular shared subs though, everyone has to use same filter for this reason.

jbertram commented:
I found some additional use-cases with JMS _queue_ consumers + FQQN + selector that weren't working as expected so I've fixed those and added a few corresponding tests.

As for the other use-cases you mentioned...I think any use-case involving a pre-created queue with a filter being accessed via FQQN by a client also defining its own filter is extremely rare if in use anywhere at all. It's certainly beyond what is documented and won't work consistently across JMS clients even now.

Generally speaking, in multicast + filtered use-cases the filtering should be done _at the queue_. In anycast + filtered use-cases the filtering should be done _at the consumer_. These basic semantics are enforced for non-FQQN use-cases and should be enforced for FQQN use-cases as well. If this breaks undocumented, untested, niche use-cases then that can be addressed in another PR if we deem it necessary.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/jms/multiprotocol/JMSFQQNConsumerTest.java:
```diff
@@ -0,0 +1,101 @@
+      final String value = RandomUtil.randomString();
+      final String filter = prop + "='" + value + "'";
+      try (Connection c = supplier.createConnection()) {
+         Session s = c.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         Queue q = s.createQueue(CompositeAddress.toFullyQualified(address, queue));
+         MessageConsumer mc = s.createConsumer(q, filter);
+         Wait.assertTrue(() -> server.locateQueue(queue) != null, 2000, 100);
+         assertNull(server.locateQueue(queue).getFilter());
+         MessageProducer p = s.createProducer(q);
+         Message m = s.createMessage();
+         m.setStringProperty(prop, value);
+         p.send(m);
+         c.start();
+         assertNotNull(mc.receive(1000));
+         m = s.createMessage();
+         m.setStringProperty(prop, RandomUtil.randomString());
+         assertNull(mc.receive(1000));
```
The second creation+prop-set for 'm' seems unused since all it does after is cehck nothing is received.

Some separation would help readability, perhaps make things like that more obvious.

rvais replied:
I personally would put this to a loop and sent several messages.

gemmellr reviewed (commented):
I wasnt speaking to explicitly defined queues only, but simply any 'existing queue' encountered. Say also, a queue already created by the current/previous buggy client and/or broker versions when trying to access them....so not necessarily a queue with a different existing filter, but perhaps even one created without any filter when it was 'meant' to have one.

Regardless how the queue comes to pre-exist, an application creating an FQQN consumer with a filter, should reasonably expect to never receive a message that doesnt match the specified filter. Even with this change the Core consumer still could, so I'd consider that still broken.

bszeti commented:
Any idea when this PR will be merged? Any chance to include in AMQ v7.11.1?

clebertsuconic commented:
It should be merged before 2.29.0 (upstream). Tomorrow probably.


We can talk internally about redhat releases then.

clebertsuconic commented:
@jbertram this still in review. but the changes seem ok.. did you run the test?

rvais reviewed (commented):
I am joining to what Robbie has commented already. There are no additional issues that I would be able to find. I think that both variants of the tests could use a loop that sends several messages. The topic variant of the test could go with couple of consumers where each filters based on different property and or value. I am open to be convinced otherwies.

rvais commented on artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQSession.java:
```diff
@@ -848,7 +848,7 @@ private ActiveMQMessageConsumer createConsumer(final ActiveMQDestination dest,
                      if (response.isAutoCreateQueues()) {
                         try {
-                           createQueue(dest, RoutingType.MULTICAST, dest.getSimpleAddress(), null, true, true, response);
+                           createQueue(dest, RoutingType.MULTICAST, dest.getSimpleAddress(), coreFilterString, true, true, response);
```
This usecase could also use a test. Now that I think about it. Wouldn't it make sense to pair the filter with session, rather than the queue? In such case, the messages that don't fit the with the filter could still be consumed by a different client later on. Especially in case the queue haven't been created with any filter previously.

gemmellr replied:
The main reason to do it on the queue is specifically to avoid messages that 'werent asked for' originally for a topic-subscription-like-FQQN-queue (note this bit is specific to multicast/topic) from accumulating on the queue where they wont then be consumed (as e.g is happening for an AMQP case before this PR).

For real-topic-subscriptions comparison, JMS doesnt allow selectors to differ on a single sub (while sharing) and changing it later constitutes a new subscription beginning from that point without previous messages.

clebertsuconic commented:
a new PR was sent to replace this one https://github.com/apache/activemq-artemis/pull/4502
